File: langchain-tutorials/1-TUI_LangGraph_agent_Tavily/tui_langgraph_agent_tavily.py
# Imports
import os
from dotenv import load_dotenv
from langchain_community.utilities.tavily_search import TavilySearchAPIWrapper
from langchain_community.tools.tavily_search import TavilySearchResults
from langchain_openai import ChatOpenAI
from langgraph.prebuilt import create_react_agent
from langchain_core.messages import HumanMessage
from rich.console import Console

# Load environment variables
load_dotenv()

# Set up rich console for output formatting
console = Console()

# Tavily API wrapper
tavily_config = TavilySearchAPIWrapper(tavily_api_key=os.getenv("TAVILY_API_KEY"))

# Initialize the language model with the OpenAI API key and model
llm = ChatOpenAI(api_key=os.getenv("OPENAI_API_KEY"), model="gpt-3.5-turbo")

# Set up Tavily search results with a maximum of 1 result
tavily_search = TavilySearchResults(api_wrapper=tavily_config, max_results=1)

# Create the agent executor with the specified model and tools
agent_executor = create_react_agent(model=llm, tools=[tavily_search])


# Define a function to process agent response
def process_chunk(chunk):
    """
    Process a chunk of data and extract relevant information.

    Args:
        chunk (dict): The chunk of data to be processed.

    Returns:
        None

    This function takes a chunk of data as input and processes it to extract relevant information. It checks if the chunk contains an "agent" key. If it does, it iterates over the messages in the "agent" dictionary. If a message contains "tool_calls" in its additional_kwargs, it extracts the tool call information. It iterates over the tool calls and extracts the tool name, tool arguments, and tool query. It then prints the tool name, tool query, and a message indicating that the tool is being called. If a message does not contain "tool_calls" in its additional_kwargs, it extracts the agent response and prints it.

    Note:
        This function assumes that the input chunk is a dictionary with the structure {"agent": {"messages": [...]}}.
    """

    if "agent" in chunk:
        for message in chunk["agent"]["messages"]:
            if "tool_calls" in message.additional_kwargs:
                # Extract tool call information
                tool_calls = message.additional_kwargs["tool_calls"]
                for tool_call in tool_calls:
                    tool_arguments = eval(tool_call["function"]["arguments"])
                    extract_tool_name = tool_call["function"]["name"]
                    extract_tool_query = tool_arguments["query"]

                    console.print(
                        f"\nThe agent is calling the tool [white on deep_sky_blue1]'{extract_tool_name}'[/white on deep_sky_blue1] with the query [white on deep_sky_blue1]'{extract_tool_query}'[/white on deep_sky_blue1].\nWait for the answer.",
                        style="deep_sky_blue1",
                    )
            else:
                # Extract agent response
                console.print(f"\nAgent:\n{message.content}", style="black on white")


# Loop until the user enters "quit"
while True:
    # Get the user question and display it in the console
    user_question = input("\nUser:\n")

    # Check if the user wants to quit
    if user_question.lower() == "quit":
        console.print("\nAgent:\nHave a nice day! :wave:\n", style="black on white")
        break

    # Run the agent on the user question
    agent_answer = agent_executor.invoke(
        {
            "messages": [
                HumanMessage(content=user_question),
            ]
        }
    )

    # Get the agent answer and display it in the console
    for chunk in agent_executor.stream(
        {"messages": [HumanMessage(content=user_question)]}
    ):
        process_chunk(chunk)

File: langgraph-tutorials/1-TUI_LangGraph_agent_Tavily/nodejs/tui_langgraph_agent_tavily.js
// Imports
import dotenv from "dotenv";
import readline from "readline";
import { ChatOpenAI } from "@langchain/openai";
import { TavilySearchResults } from "@langchain/community/tools/tavily_search";
import { createReactAgent } from "@langchain/langgraph/prebuilt";
import { HumanMessage } from "@langchain/core/messages";
import term from "terminal-kit";

// Initialize dotenv to load environment variables
dotenv.config();

// Initialize Terminal Kit for better output formatting and visualization
const tk = term.terminal;

// Initialize OpenAI LLM
const llm = new ChatOpenAI({
  apiKey: process.env.OPENAI_API_KEY,
  modelName: "gpt-3.5-turbo",
});

// Initialize Tavily
const tavily = new TavilySearchResults({
  maxResults: 3,
});

// Create a LangGraph agent
const langgraphAgent = createReactAgent({
  llm: llm,
  tools: [tavily],
});

// Define an asynchronous function to get the user question
function getUserQuestion(message) {
  // Creating a readline interface for reading lines from the standard input (keyboard)
  const rl = readline.createInterface({
    input: process.stdin, // Setting the input stream to the standard input (keyboard)
    output: process.stdout, // Setting the output stream to the standard output (console)
  });

  // Returning a Promise that resolves when the user enters something
  return new Promise((resolve) => {
    // Asking the user for input with the provided message
    rl.question(message, (userQuestion) => {
      // Closing the readline interface after receiving input
      rl.close();
      // Resolving the Promise with the user question
      resolve(userQuestion);
    });
  });
}

// Define a function to process chunks from the agent
function processChunks(chunk) {
  /**
   * Processes a chunk from the agent and displays information about tool calls or the agent's answer.
   *
   * @param {Object} chunk - The chunk to be processed.
   * @return {void}
   */

  // Check if the chunk contains an agent's message
  if ("agent" in chunk) {
    // Iterate over the messages in the chunk
    for (const message of chunk.agent.messages) {
      // Check if the message contains tool calls
      if (
        "tool_calls" in message.additional_kwargs != undefined &&
        Array.isArray(message.additional_kwargs.tool_calls)
      ) {
        // If the message contains tool calls, extract and display an informative message with tool call details

        // Extract all the tool calls
        const toolCalls = message.additional_kwargs.tool_calls;

        // Iterate over the tool calls
        toolCalls.forEach((toolCall) => {
          // Extract the tool name
          const toolName = toolCall.function.name;

          // Extract the tool input
          const toolArguments = JSON.parse(
            toolCall.function.arguments.replace(/'/g, '"')
          );
          const toolInput = toolArguments.input;

          // Display an informative message with tool call details
          tk
            .colorRgbHex("#00afff")(`\nThe agent is calling the tool `)
            .bgColorRgbHex("#00afff")
            .color("black")(`${toolName}`)
            .bgColor("black")
            .colorRgbHex("#00afff")(` with the query `)
            .bgColorRgbHex("#00afff")
            .color("black")(`${toolInput}`)
            .bgColor("black")
            .colorRgbHex("#00afff")(
            `. Please wait for the agent's answer...\n`
          );
        });
      } else {
        // If the message doesn't contain tool calls, extract and display the agent's answer

        // Extract the agent's answer
        const agentAnswer = message.content;

        // Display the agent's answer
        tk.bgColor("white")
          .color("black")(`\nAgent:\n${agentAnswer}\n`)
          .color("white")
          .bgColor("black");
      }
    }
  }
}

// Define the main function
async function main() {
  /**
   * Runs the main loop of the chat application.
   *
   * @return {Promise<void>} A promise that resolves when the user chooses to quit the chat.
   */

  // Loop until the user chooses to quit the chat
  while (true) {
    // Get the user's question and display it in the terminal
    const userQuestion = await getUserQuestion("\nUser:\n");

    // Check if the user wants to quit the chat
    if (userQuestion.toLowerCase() === "quit") {
      tk.bgColor("white").color("black")("\nAgent:\nHave a nice day!\n");
      tk.bgColor("black").color("white")("\n");
      break;
    }

    // Use the stream method of the LangGraph agent to get the agent's answer
    const agentAnswer = await langgraphAgent.stream({
      messages: [new HumanMessage({ content: userQuestion })],
    });

    // Process the chunks from the agent
    for await (const chunk of agentAnswer) {
      processChunks(chunk);
    }
  }
}

// Call the main function
main();
